librz/bp/serialize_bp.c

```diff
@@ -0,0 +1,326 @@
+		case BP_FIELD_NAME:
+			if (child->type != RZ_JSON_STRING) {
+				break;
+			}
+			bp_item_temp.name = (char *)child->str_value;
+			break;
+		case BP_FIELD_MODULE_NAME:
+			if (child->type != RZ_JSON_STRING) {
+				break;
+			}
+			bp_item_temp.module_name = (char *)child->str_value;
+			break;
+		case BP_FIELD_MODULE_DELTA:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			bp_item_temp.module_delta = child->num.s_value;
+			break;
+		case BP_FIELD_DELTA:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			bp_item_temp.delta = child->num.u_value;
+			break;
+		case BP_FIELD_SIZE:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			bp_item_temp.size = (int)child->num.s_value;
+			break;
+		case BP_FIELD_SWSTEP:
+			if (child->type != RZ_JSON_BOOLEAN) {
+				break;
+			}
+			bp_item_temp.swstep = child->num.u_value ? true : false;
+			break;
+		case BP_FIELD_PERM:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			bp_item_temp.perm = (int)child->num.s_value;
+			break;
+		case BP_FIELD_HW:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			bp_item_temp.hw = (int)child->num.s_value;
+			break;
+		case BP_FIELD_TRACE:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			bp_item_temp.trace = (int)child->num.s_value;
+			break;
+		case BP_FIELD_INTERNAL:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			bp_item_temp.internal = (int)child->num.s_value;
+			break;
+		case BP_FIELD_ENABLED:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			bp_item_temp.enabled = (int)child->num.s_value;
+			break;
+		case BP_FIELD_TOGGLEHITS:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			bp_item_temp.togglehits = (int)child->num.s_value;
+			break;
+		case BP_FIELD_HITS:
+			if (child->type != RZ_JSON_INTEGER) {
+				break;
+			}
+			bp_item_temp.hits = (int)child->num.s_value;
+			break;
+		case BP_FIELD_PIDS:
+			if (child->type != RZ_JSON_ARRAY) {
+				break;
+			}
+			int index = 0;
+			for (const RzJson *pid_child = child->children.first; pid_child; pid_child = pid_child->next) {
+				if (index >= RZ_BP_MAXPIDS) {
+					break;
+				}
+				bp_item_temp.pids[index] = (int)pid_child->num.s_value;
+				++index;
+			}
+			break;
+		case BP_FIELD_DATA:
+			if (child->type != RZ_JSON_STRING) {
+				break;
+			}
+			bp_item_temp.data = (char *)child->str_value;
+			break;
+		case BP_FIELD_COND:
+			if (child->type != RZ_JSON_STRING) {
+				break;
+			}
+			bp_item_temp.cond = (char *)child->str_value;
+			break;
+		case BP_FIELD_EXPR:
+			if (child->type != RZ_JSON_STRING) {
+				break;
+			}
+			bp_item_temp.expr = (char *)child->str_value;
+			break;
+	})
+
+	RzBreakpointItem *bp_item = NULL;
+	if (bp_item_temp.hw) {
+		bp_item = rz_bp_add_hw(ctx->bp, bp_item_temp.addr, bp_item_temp.size, bp_item_temp.perm);
+	} else {
+		bp_item = rz_bp_add_sw(ctx->bp, bp_item_temp.addr, bp_item_temp.size, bp_item_temp.perm);
+	}
+	if (!bp_item) {
+		goto beach;
+	}
```

Comment:
I was testing this on Windows with:
```
λ rizin-install\bin\rizin.exe -p rizin.rzdb -d rizin-install\bin\rizin.exe
DebugActiveProcess: (0x5) Access is denied.
DebugActiveProcess: (0x5) Access is denied.
WARNING: Breakpoint won't be placed since it's not in a valid map.
You can bypass this check by setting dbg.bpinmaps to false.
Spawned new process with pid 22532, tid = 21028
```
The breakpoints seem to work then, but I do wonder whether the restoring of the debug info happens at the right time... @thestr4ng3r idea? Is it alright as a first step?

Also, what about PIE? This is not going to work, is it?

Reply:
Thanks for testing on Windows. I'll check whether this works for PIE or not.

Reply:
The serialization works because (it seems) that base addresses are (maybe) stored in the project files. This seems to work for both PIE and no-PIE.

Reply:
It is probably alright that it doesn't work, because IO/Bin is not properly serialized yet... let's see.

Reply:
Right now, before IO/Bin serialization is implemented, the only available information is a filename. No base address or anything similar. But even with IO and Bin, that still won't give away much info about mapping in a debug process. That is a whole different story.

The idea would be to never load or save a project in debug mode, but only in static mode and then switch to debug with `ood`. Then the regular rebasing that happens should take care of relocating the breakpoints to wherever the file is statically loaded by default.